eyelid-match-ops/src/primitives/yashe.rs

```diff
@@ -0,0 +1,119 @@
+//! Implementation of YASHE cryptosystem
+//! `<https://eprint.iacr.org/2013/075.pdf>`
+
+use crate::primitives::poly::{inverse, modular_poly::Poly, Coeff};
+use ark_ff::{One, UniformRand, Zero};
+use rand::rngs::ThreadRng;
+use rand_distr::{Distribution, Normal};
+
+pub mod test;
+
+/// Yashe parameters
+#[derive(Clone)]
+pub struct YasheParams {
+    /// Plaintext coefficient modulus
+    pub t: u64,
+    /// Standard deviation
+    pub delta: f64,
+}
+
+/// Yashe scheme
+pub struct Yashe<const MAX_POLY_DEGREE: usize> {
+    /// Cryptosystem parameters
+    params: YasheParams,
+}
+
+/// Private key struct
+#[derive(Debug, Clone)]
+#[allow(dead_code)]
+pub struct PrivateKey<const MAX_POLY_DEGREE: usize> {
+    /// Sampled with small coefficients (and invertible)
+    f: Poly<MAX_POLY_DEGREE>,
+    /// The inverse of f
+    finv: Poly<MAX_POLY_DEGREE>,
+    /// Private key
+    priv_key: Poly<MAX_POLY_DEGREE>,
+}
+
+/// Public key struct
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct PublicKey<const MAX_POLY_DEGREE: usize> {
+    /// Public key
+    h: Poly<MAX_POLY_DEGREE>,
+}
+
+impl<const MAX_POLY_DEGREE: usize> Yashe<MAX_POLY_DEGREE> {
+    /// Yashe constructor
+    pub fn new(params: YasheParams) -> Self {
+        Self { params }
+    }
+
+    /// Generate the private key
+    pub fn generate_private_key(&self, rng: ThreadRng) -> PrivateKey<MAX_POLY_DEGREE> {
+        loop {
+            let f = self.sample_gaussian(rng.clone());
+            let mut priv_key = f.clone();
+            let finv = inverse(&f);
+
+            priv_key *= Coeff::from(self.params.t);
+            priv_key[0] += Coeff::one();
+            priv_key.truncate_to_canonical_form();
+
+            if finv.is_ok() {
+                let finv = finv.unwrap();
```

Comment:
Rust has a special syntax for this which is more performant than calling the potentially panicking function `unwrap()`:

```suggestion
            if let Ok(finv) = finv {
```